aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -0,0 +1,154 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.s3;
+
+import com.amazonaws.services.s3.AmazonS3URI;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.core.sync.ResponseTransformer;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+
+public class S3InputStream extends SeekableInputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(S3InputStream.class);
+
+  private final StackTraceElement[] createStack;
+  private final S3Client s3;
+  private final AmazonS3URI location;
+
+  private InputStream stream;
+  private long pos = 0;
+  private long next = 0;
+  private boolean closed = false;
+
+  private int skipSize = 1024 * 1024;
+
+  public S3InputStream(S3Client s3, AmazonS3URI location) {
+    this.s3 = s3;
+    this.location = location;
+
+    createStack = Thread.currentThread().getStackTrace();
+  }
+
+  @Override
+  public long getPos() {
+    return pos;
+  }
+
+  @Override
+  public void seek(long newPos) {
+    Preconditions.checkState(!closed, "already closed");
+    Preconditions.checkArgument(newPos >= 0, "position is negative: %s", newPos);
+
+    // this allows a seek beyond the end of the stream but the next read will fail
+    next = newPos;
+  }
+
+  @Override
+  public int read() throws IOException {
+    return stream.read();
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    Preconditions.checkState(!closed, "already closed");
+    positionStream();
+
+    int bytesRead = stream.read(b, off, len);
+    pos += bytesRead;
+    next += bytesRead;
+
+    return bytesRead;
+  }
+
+  @Override
+  public void close() throws IOException {
+    super.close();
+    closed = true;
+    closeStream();
+  }
+
+  private void positionStream() throws IOException {
+    if ((stream != null) && (next == pos)) {
+      // already at specified position
+      return;
+    }
+
+    if ((stream != null) && (next > pos)) {
+      // seeking forwards
+      long skip = next - pos;
+      if (skip <= Math.max(stream.available(), skipSize)) {
+        // already buffered or seek is small enough
+        LOG.debug("Read-through seek for {} to offset {}", location, next);
+        try {
+          if (stream.skip(skip) == skip) {
```

Comment:
`InputStream.skip` doesn't guarantee that the bytes skipped is equal to the request: "The skip method may, for a variety of reasons, end up skipping over some smaller number of bytes, possibly 0."

In this case, if anything other than the expected number of bytes is skipped, then this will fall back to reconnecting.

I don't think that's the desired behavior because S3 may already be sending bytes to the reader (and it will be faster to read through) but the skip returns early because not as many bytes as requested were immediately available. I think that this should probably use a `skipFully` method that loops and skips until there are no more bytes to read through.

Reply:
I actually agree with this an flagged it when I saw what presto was doing.  I believe the Presto implementation relies skip skipping correctly or it just falls back to closing the stream and opening a new one (which seems really wasteful).

I'll update this to skip fully.